### ISSUE TITLE
Clarification on ReqRep's NNG_OPT_REQ_RESENDTIME behaviour

### ISSUE THREAD

**Issue by AlexKornitzer** [nanomsg/nng#1240]
Apologies in advance as this is most likely a misunderstanding on my part, but reading the descriptions for ReqRep and NNG_OPT_REQ_RESENDTIME leave me a tad unclear on certain intricacies of the expected behaviour.

From my understanding of the documentation if I have a client that sends message to a server in a loop, but the server takes longer than the resend time to respond, the client will send the message again. Now assuming we had a timeout of 1 second and we make the server respond after 5 seconds, give or take, we should end up with the server being sent five messages. Since the client is in a loop when we hit send the requester will now ignore any reply from the earlier request, but the client has to wait for the server to handle all 5 messages from the previous request. Is it the case here that for nng the 5 messages are actually all part of the earlier request and waiting for all of them to be processed is intended?

## NNG & Platform details.

nng: git 1a00f65
os: arch linux
lang: c (reqrep example code - modified)

## Expected Behavior

When connecting to the server with a client that sends and receives a messages in a loop, the new send call should cause the server to discard any pending messages from those caused by a resend timeout as we can deduce that a new send on the socket means the client received the reply.

## Actual Behavior

What actually happens is the client sits waiting on the recv for the new send and will not get a response until the server has consumed the resend back log.

## Steps to Reproduce
https://gist.github.com/AlexKornitzer/371eef55e558e89d10850d53986dfb35

```
./reqrep server ipc:///tmp/abcd 
./reqrep client ipc:///tmp/abcd
```

## Additionally

On the latest build of nng if I start and kill the client a couple of times I get the following crash in the server:
```
panic: pthread_mutex_lock: Invalid argument
This message is indicative of a BUG.
Report this at https://github.com/nanomsg/nng/issues
/home/developer/Developer/scratch/nng/reqrep(+0x15bb9) [0x555555569bb9]
/home/developer/Developer/scratch/nng/reqrep(+0x1f7af) [0x5555555737af]
/home/developer/Developer/scratch/nng/reqrep(+0x1f958) [0x555555573958]
/home/developer/Developer/scratch/nng/reqrep(+0x1c826) [0x555555570826]
/home/developer/Developer/scratch/nng/reqrep(+0x1c562) [0x555555570562]
/home/developer/Developer/scratch/nng/reqrep(+0xf21f) [0x55555556321f]
/home/developer/Developer/scratch/nng/reqrep(+0x24617) [0x555555578617]
/home/developer/Developer/scratch/nng/reqrep(+0x15f4e) [0x555555569f4e]
/home/developer/Developer/scratch/nng/reqrep(+0xe60d) [0x55555556260d]
/home/developer/Developer/scratch/nng/reqrep(+0xec35) [0x555555562c35]
/home/developer/Developer/scratch/nng/reqrep(+0x18a00) [0x55555556ca00]
/home/developer/Developer/scratch/nng/reqrep(+0x739f) [0x55555555b39f]
/home/developer/Developer/scratch/nng/reqrep(+0x710d) [0x55555555b10d]
/home/developer/Developer/scratch/nng/reqrep(+0x6f29) [0x55555555af29]
/home/developer/Developer/scratch/nng/reqrep(+0x690f) [0x55555555a90f]
/home/developer/Developer/scratch/nng/reqrep(+0x6dac) [0x55555555adac]
/usr/lib/libc.so.6(__libc_start_main+0xf3) [0x7ffff7ded023]
/home/developer/Developer/scratch/nng/reqrep(+0x672e) [0x55555555a72e]

(gdb) bt
#0  0x00007ffff7e01ce5 in raise () from /usr/lib/libc.so.6
#1  0x00007ffff7deb857 in abort () from /usr/lib/libc.so.6
#2  0x00005555555733bf in nni_plat_abort ()
#3  0x0000555555569bbe in nni_panic ()
#4  0x00005555555737af in nni_pthread_mutex_lock ()
#5  0x0000555555573958 in nni_plat_mtx_lock ()
#6  0x0000555555570826 in nni_mtx_lock ()
#7  0x0000555555570562 in nni_task_dispatch ()
#8  0x000055555556321f in nni_aio_begin ()
#9  0x0000555555578617 in ipctran_pipe_recv ()
#10 0x0000555555569f4e in nni_pipe_recv ()
#11 0x000055555556260d in rep0_ctx_recv ()
#12 0x0000555555562c35 in rep0_sock_recv ()
#13 0x000055555556ca00 in nni_sock_recv ()
#14 0x000055555555b39f in nng_recv_aio ()
#15 0x000055555555b10d in nng_recvmsg ()
#16 0x000055555555af29 in nng_recv ()
#17 0x000055555555a90f in server ()
#18 0x000055555555adac in main ()
```

**Comment by gdamore** [comment]
The panic you're seeing is disconcerting.  I will investigate.

As for the resend.

The messages are replayed by the client if it does not receive a reply within the time requested.  (It also will replay if the server disconnects for some reason.)

The server may see these as separate requests, in which case, yes, the work will be duplicated if it is a big job.  It's a good idea to use some kind of sequence ID or something like that so that servers don't waste effort on expensive jobs.  (It is also critically important the the requests be idempotent.  A sequence number can help make non-idempotent requests idempotent.)

Generally, I would not recommend setting the RESEND time to anything shorter than the reasonable upper limit on the cost to perform the work.

If you send a new job, the client will forget about the old one, and discard any replies (as well as stop sending any retries of the old one), but there is no notification to the server.  There is no cancellation in the protocol itself.

Hopefully that answers some of your questions.

**Comment by AlexKornitzer** [comment]
Thank you for taking the time to respond, that does clarify things greatly. This is a great little library, I recommend it to everyone.

I have just tried to replicate the bug in rust (as this is where I have been getting a SIGSEV) and it seems to be the same, yet the reported backtrace is different, it seems to imply that we are receiving on a null pipe. This is a guess but it might be that the pipe is closed but it is still trying to handle the resend request, as I believe this only happens when resends have occurred.

```
Thread 1 "scratch" received signal SIGSEGV, Segmentation fault.
0x0000555555577ce4 in nni_pipe_recv (p=0x0, aio=0x7fffe8001430)
    at /home/developer/.cargo/registry/src/github.com-1ecc6299db9ec823/nng-sys-1.2.4-rc.1/nng/src/core/pipe.c:141
141             p->p_tran_ops.p_recv(p->p_tran_data, aio);
(gdb) bt
#0  0x0000555555577ce4 in nni_pipe_recv (p=0x0, aio=0x7fffe8001430)
    at /home/developer/.cargo/registry/src/github.com-1ecc6299db9ec823/nng-sys-1.2.4-rc.1/nng/src/core/pipe.c:141
#1  0x000055555556dddf in rep0_ctx_recv (arg=0x5555555ead20, aio=0x5555555edcb0)
    at /home/developer/.cargo/registry/src/github.com-1ecc6299db9ec823/nng-sys-1.2.4-rc.1/nng/src/protocol/reqrep0/rep.c:504
#2  0x000055555556e361 in rep0_sock_recv (arg=0x5555555eac10, aio=0x5555555edcb0)
    at /home/developer/.cargo/registry/src/github.com-1ecc6299db9ec823/nng-sys-1.2.4-rc.1/nng/src/protocol/reqrep0/rep.c:670
#3  0x000055555557ade3 in nni_sock_recv (sock=0x5555555ea220, aio=0x5555555edcb0)
    at /home/developer/.cargo/registry/src/github.com-1ecc6299db9ec823/nng-sys-1.2.4-rc.1/nng/src/core/socket.c:847
#4  0x00005555555666e6 in nng_recv_aio (s=..., aio=0x5555555edcb0)
    at /home/developer/.cargo/registry/src/github.com-1ecc6299db9ec823/nng-sys-1.2.4-rc.1/nng/src/nng.c:214
#5  0x000055555556651d in nng_recvmsg (s=..., msgp=0x7fffffffcf50, flags=0)
    at /home/developer/.cargo/registry/src/github.com-1ecc6299db9ec823/nng-sys-1.2.4-rc.1/nng/src/nng.c:136
#6  0x0000555555564a84 in nng::socket::Socket::recv (self=0x7fffffffd018)
    at /home/developer/.cargo/git/checkouts/nng-rs-585a5382ec72d3dc/fc301c2/src/socket.rs:245
#7  0x00005555555600f1 in scratch::server () at src/main.rs:76
#8  0x0000555555560339 in scratch::main () at src/main.rs:87
```

Would you prefer me to open this in a new issue that tracks this potential bug?

**Comment by gdamore** [comment]
Yes please. Open a new issue.  It’s definitely a bug.

Get Outlook for iOS<https://aka.ms/o0ukef>
________________________________
